File: bin/mish/mish.c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <string.h>
#include <unistd.h>

#define LINELEN 80
#define NARGS 10

static int readline(char *line, size_t size) {
  char *s = fgets(line, size, stdin);
  if (!s)
    return 0;
  size_t n = strlen(s);
  if (line[n-1] == '\n')
    line[--n] = '\0';
  return 1;
}

static int do_exit(int argc, char *argv[], char *envp[]) {
  exit(EXIT_FAILURE);
}

static int do_env(int argc, char *argv[], char *envp[]) {
  while (*envp)
    puts(*envp++);
  return 0;
}

typedef int (*command_t)(int, char *[], char *[]);
typedef struct { char *name; command_t cmd; } builtin_t;

static const builtin_t builtins[] = {
  { "exit", do_exit },
  { "env", do_env },
  { NULL, NULL },
};

int main(int argc, char *argv[], char *envp[]) {
  char line[LINELEN];
  char *new_argv[NARGS+1];
  int status;

  while (1) {
    write(1, "# ", 2);
    if (!readline(line, LINELEN))
      break;

    int new_argc;
    char *brkt, *word;
    for (new_argc = 0, word = strtok_r(line, " ", &brkt);
         new_argc < NARGS && word;
         new_argc++, word = strtok_r(NULL, " ", &brkt))
    {
      new_argv[new_argc] = word;
    }
    new_argv[new_argc] = NULL;

    /* Lookup builtin command */
    const builtin_t *b = NULL;
    for (b = builtins; b->name; b++)
      if (strcmp(new_argv[0], b->name) == 0)
        break;
    if (b->name) {
      status = b->cmd(new_argc, new_argv, envp);
      if (status)
        break;
      continue;
    }

    /* Process external command */
    pid_t pid;
    if ((pid = fork())) {
      waitpid(pid, &status, 0);
      if (status)
        printf("Child exited with %d\n", status);
    } else {
      fprintf(stderr, "run %s\n", line);
      execve(new_argv[0], new_argv, envp);
      perror("execve failed");
      return EXIT_FAILURE;
    }
  }

  return status;
}

File: bin/mish/mish.lua
#!/bin/lua

builtin = {}

function builtin.exit(args)
  os.exit(tonumber(args[2]))
end

internal = {}

function internal.cat(args)
  for i = 2, #args do
    for line in io.lines(args[i]) do
      print(line)
    end
  end
end

-- spawn a process
function spawn(prog, argv)
  pid = unix.fork()
  if pid then
    _, status = unix.waitpid(pid, 0)
    signo = WTERMSIG(status)
    code = WEXITSTATUS(status)
    if signo > 0 then
      print(argv[1] .. ": terminated by " .. signo .. " signal!")
    elseif code > 0 then
      print(argv[1] .. ": exited with " .. code .. " code!")
    end
  else
    if type(prog) == "function" then
      os.exit(prog(argv))
    else
      argv[1] = prog
      err = unix.execve(prog, argv)
      fail(prog .. ": " .. err.msg)
    end
  end
end

function execute(argv)
  -- lookup builtin command
  cmd = builtin[argv[1]]
  if cmd then return cmd(argv) end

  -- is that an internal command ?
  prog = internal[argv[1]]
  if prog then return spawn(prog, argv) end

  -- is that an external command ?
  prog = which(argv[1])
  if prog then return spawn(prog, argv) end

  print(argv[1] .. ": No such command")
end

function mish()
  while true do
    -- print prompt
    io.stdout:write("# ")
    io.stdout:flush()

    -- read line and split it into words
    line = io.stdin:read("l")
    if not line then break end
    argv = {}
    for word in line:gmatch("%g+") do
      table.insert(argv, word)
    end

    if #argv then pcall(execute, argv) end
  end
end

-- setup default environment variables
if not environ["PATH"] then
  environ["PATH"] = "/usr/bin:/bin"
end

mish()
